### DISCUSSION TITLE
Returning the same promise twice from an atom getter causes a Promise chaining cycle.

### DISCUSSION THREAD

**Discussion by iwoplaza** [pmndrs/jotai#2151] [Bug report] [answered]
## Context
**Tested on `jotai: 2.4.2`, but seems to also occur in versions between 2.1.x - 2.4.2**

## Issue
After migrating to a newer version of Jotai, we noticed these errors popping up in the console.

<img width="588" alt="image" src="https://github.com/pmndrs/jotai/assets/7166752/d550c841-2e01-400d-b8a1-02affc217a44">

This error can be reproduced in vanilla js with a couple lines of code. It essentially means that a promise was resolved with itself, so it can never resolve.

<img width="575" alt="image" src="https://github.com/pmndrs/jotai/assets/7166752/dce36985-9d67-41fd-92d3-7591dda07a84">

## Reproduction

In our project's codebase, we frequently wrap an asynchronous atom and provide additional sync dependencies, so that we can for example conditionally return the Promise, or an actual concrete value to reduce unnecessary Suspensions.

Having an async atom:
```ts
const dataAtom = atom(async (get) => {
  // Async data, such as fetching from an endpoint or interacting with a Web API.
});
```

We wrap it with an atom that just returns the result of the atom above, but adds more dependencies:
```ts
const derivedAtom = atom((get) => {
  get(counterAtom); // sync dependency

  return get(dataAtom); // still returns a promise, so this is also an async atom
});
```

The derived atom is invalidated once `counterAtom` changes, but it still returns the same promise. While looking through the source-code, I stumbled upon the potential pain point:
```js
  const setAtomState = (atom, atomState) => {
    if ((import.meta.env ? import.meta.env.MODE : void 0) !== "production") {
      Object.freeze(atomState);
    }
    const prevAtomState = atomStateMap.get(atom);
    atomStateMap.set(atom, atomState);
    if (!pendingMap.has(atom)) {
      pendingMap.set(atom, prevAtomState);
    }
    if (prevAtomState && hasPromiseAtomValue(prevAtomState)) {
      const next = "v" in atomState ? atomState.v instanceof Promise ? atomState.v : Promise.resolve(atomState.v) : Promise.reject(atomState.e);
      // HERE \/ - we cancel the promise even though `next` could equal `prevAtomState.v`, which causes the chaining cycle.
      cancelPromise(prevAtomState.v, next);
    }
  };
```

If we do not check it there, then the promise will get resolved with itself down the line:
```js
  const setAtomValueOrPromise = (atom, valueOrPromise, nextDependencies, abortPromise) => {
    if (isPromiseLike(valueOrPromise)) {
      let continuePromise;
      const promise = new Promise((resolve, reject) => {
      
        // ...
        
        continuePromise = (next) => {
          if (!settled) {
            settled = true;
            next.then(
              (v) => resolvePromise(promise, v),
              (e) => rejectPromise(promise, e)
            );
            resolve(next); // <-- HERE
          }
        };
        
        // ...
  };
```

***Reproduction Sandbox:***
https://codesandbox.io/s/jotai-returning-same-promise-twice-from-atom-getter-8x7nt4

**Comment by dai-shi** [comment]
Thanks for reporting.
Thought #1851 caused this, but it's working.
Now, I wonder which PR changed the behavior. If you find anything, feel free to share it.

**Reply by dai-shi** [reply]
> Tested on jotai: 2.4.2, but seems to also occur in versions between 2.1.x - 2.4.2

Oh, actually v2.1.0 works.

**Reply by dai-shi** [reply]
It's #1933 that changed the behavior: https://codesandbox.io/s/jotai-returning-same-promise-twice-from-atom-getter-forked-z4pdzl
Yeah, it makes sense.

**Comment by dai-shi** [comment]
> // HERE \/ - we cancel the promise even though `next` could equal `prevAtomState.v`, which causes the chaining cycle.

So, does this fix it?
```js
if (prevAtomState.v !== next) {
  cancelPromise(prevAtomState.v, next);
}
```

**Reply by iwoplaza** [reply]
That was my guess, but I haven’t yet implemented it myself. I will probably fork and create a PR for it in a couple minutes.

**Reply by dai-shi** [reply, marked as answer]
https://github.com/pmndrs/jotai/discussions/2151#discussioncomment-7122397 I'm still interested in what made this change.
We also need to create a failing test case for it, preferably in advance.

(Looking forward to your contribution!)
